libraries/eosiolib/contracts/eosio/privileged.hpp

```diff
@@ -185,6 +212,203 @@ namespace eosio {
     */
    void get_blockchain_parameters(eosio::blockchain_parameters& params);
 
+   /**
+    *  Helper structure for host functions set_parameters_packed and get_parameters_packed
+    *  @ingroup privileged
+    */
+   struct params_object{
+      std::string packed;
+
+      template<typename T>
+      params_object(T v){
+         (*this)(v);
+      }
+
+      params_object(const std::string& params) : packed(params) {}
+
+      bool operator == (const params_object& pp) const{
+         return packed == pp.packed;
+      }
+
+      params_object& operator () (unsigned_int val){
+         char buffer[4];
+         datastream<char*> ds((char*)&buffer, sizeof(buffer));
+         ds << val;
+         ASSERT_LE(ds.tellp(), sizeof(buffer));
+
+         packed += {buffer, ds.tellp()};
+         return *this;
+      }
+
+      params_object& operator () (uint16_t val){
+         char buffer[2];
+         datastream<char*> ds((char*)&buffer, sizeof(buffer));
+         ds << val;
+         ASSERT_EQ(ds.tellp(), sizeof(buffer));
+
+         packed += {buffer, ds.tellp()};
+         return *this;
+      }
+
+      params_object& operator () (uint32_t val){
+         char buffer[4];
+         datastream<char*> ds((char*)&buffer, sizeof(buffer));
+         ds << val;
+         ASSERT_EQ(ds.tellp(), sizeof(buffer));
+
+         packed += {buffer, ds.tellp()};
+         return *this;
+      }
+
+      params_object& operator () (uint64_t val){
+         char buffer[8];
+         datastream<char*> ds((char*)&buffer, sizeof(buffer));
+         ds << val;
+         ASSERT_EQ(ds.tellp(), sizeof(buffer));
+
+         packed += {buffer, ds.tellp()};
+         return *this;
+      }
+
+      params_object& operator () (std::vector<unsigned_int>&& val){
+         char buffer[512];
+         datastream<char*> ds((char*)&buffer, sizeof(buffer));
+         ds << val;
+         ASSERT_LESS(ds.tellp(), sizeof(buffer));
+
+         packed += {buffer, ds.tellp()};
+         return *this;
+      }
+
+      void set() const{
```

Comment:
Is it possible to refactor struct `params_object` and add functions e.g. `set_parameters_packed(const eosio::params_object& params)` (and similar for the getter) outside of this struct?  (It's mentioned there is obstacle in doing so, please remind me what the difficulty is 🤔)

Reply:
I think it is possible, i copied this params_object directly from the eos example (linked in the description of the PR), do you think it would be better to put it on this way?

Reply:
Since other similar functions are implemented on this other way (e.g. "set/get_wasm_parameters" uses struct `wasm_parameters`, and "set/get_blockchain_parameters" with `blockchain_parameters`), it seems to me more consistent in API design if users can call functions like "set/get_parameters_packed" with `params_object` as parameter. Please let me know your thoughts:)

Reply:
Yes, i agree to have this consistency in the way the API is called, i will make the changes.

Reply:
Hi Victor,  I saw you packed  eosio::params_object as parameter of set_parameters_packed,  but however please follow the design document https://developers.eos.io/manuals/eos/v2.1/classeosio_1_1chain_1_1webassembly_1_1interface#function-get_blockchain_parameters_packed, that says: a packed data stream which represents an encoded sequence of parameter_id:paramter_value pairs with the following format: |varuint32:sequence_length | varuint32:parameter_id | :parameter_value | ... .   I know the eos part didn't strictly follow the design document, but this is a change to make it correct.

Reply:
And also the eosio::params_object is not efficient at all.  Using a signle datastream and 1 buffer to pack all data is efficient. but if using  each 1 data using a new buffer and new datastream, and then try move it to a string, that is not efficient.

And also  eosio::params_object  as parameter is hard to use by cleos input, did you ever think that?

Reply:
I removed this code, as the get_parameters_packed and set_parameters packed host functions were implemented in a different way.